### PR TITLE
Add Shape and Text json schemas for Mobile Session Replay

### PR DESCRIPTION
**Pull request by mariusc83**
In this PR we are introducing the JSON schemas for the Mobile Session Replay feature. 
The related RFC based on which the schemas were defined can be found [here](https://datadoghq.atlassian.net/wiki/spaces/RUMP/pages/2471460887/RFC+-+Data+Format+for+Mobile+Session+Replay)

For now we only introduced the JSON schemas for the `ShapeWireframe` and `TextWireframe` but more will come later.

### CODE REVIEW

File: validate.sh
Comment (mariusc83) [inline validate.sh:10]: Could not find a way to use `Map<String,String>` in bash so had to stick with arrays :)

**Comment by xgouchet** [comment]
A couple of question, why add a new level in the samples folder hierarchy for `rum-schemas` vs `session-replay-schemas`, when there are already folders for  `rum` vs `telemetry`? 
Also maybe we should add `@DataDog/rum-session-replay` team as a code owner here

**Comment by mariusc83** [comment]
> A couple of question, why add a new level in the samples folder hierarchy for `rum-schemas` vs `session-replay-schemas`, when there are already folders for `rum` vs `telemetry`? Also maybe we should add `@DataDog/rum-session-replay` team as a code owner here

@xgouchet Because the `rum` and `telemetry` folders are using the RUM schema format for validation, `session-replay` instead is using a different schema so I needed a way to group those for being able to make it work with the `validation.sh` script.
Regarding the code owners part is good idea, I will add this now.

File: schemas/session-replay/shape-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/shape-wireframe-schema.json:16]: The `"type"` attribute should be `required` in all concrete wireframes, so here and in `text-wireframe-schema.json`.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/_common-shape-wireframe-schema.json]: Why making these two `required`? IMO:
- both have obvious default value for the player (`opacity: 1`, `cornerRadius: 0`);
- in practise majority of wireframes we record will use these defaults.

Making it required will only bloat the payloads in most cases where the wireframe uses default values. By making it optional we will safe 2 attributes per WF in majority of records.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/_common-shape-wireframe-schema.json]: Don't we think both are an aspect of the `shapeStyle`? IMO it's reasonable to make it an attribute of `shapeStyle`.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/_common-shape-wireframe-schema.json]: Similar, IMO nothing in there should be `required` and all properties can be optional as we can find sensible defaults. In case of `fillColor`, the default can be "transparent".

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/_common-shape-wireframe-schema.json]: > "The filling color for this wireframe."

Here we miss an information on the color format. I guess we assume hex string, but it's worth being explicit.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/_common-shape-wireframe-schema.json]: I wonder if we're going to use `#RRGGBB` or `#RRGGBBAA` or allow both? Whereas it can be less complication for the player, mind that we need a custom implementation in recorders (iOS doesn't operate with hex colors and it uses R, G, B directly). It's IMO worth being more explicit.

File: schemas/session-replay/text-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/text-wireframe-schema.json]: Don't we also want to make `textStyle` required? It will be hard for the player to find sensible defaults to all text properties (including font size).

File: schemas/session-replay/text-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/text-wireframe-schema.json]: /nit - it's not only reserved for text fields. It can be used for "labels", "text areas", "segmented control titles" and many other platform-specific items - so better to stay platform agnostic.
```suggestion
          "description": "The style of the text.",
```

File: schemas/session-replay/text-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/text-wireframe-schema.json]: I believe it should be just `"size"` not `"font"`.

File: schemas/session-replay/text-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/text-wireframe-schema.json]: Similar - we should be more specific on the color format (`#RRGGBB` or `#RRGGBBAA`).

File: schemas/session-replay/text-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/text-wireframe-schema.json]: Similar - both seem to have good defaults: `TL` + `0, 0, 0, 0` which will cover majority of use cases. IMO no reason to send it for every text wireframe.

File: schemas/session-replay/text-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/text-wireframe-schema.json]: Similar - each defaults to `0`, so we can avoid the noise if of sending it.

File: schemas/session-replay/text-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/text-wireframe-schema.json]: Similar - as we have good defaults to `0`, we can make it optional.

File: schemas/session-replay/text-wireframe-schema.json
Comment (ncreated) [inline schemas/session-replay/text-wireframe-schema.json:38]: There was a good idea (i.a. from @louiszawadzki) to also include font type property which will help with finding a backup font (e.g. "serif" or "sans-serif").

**Review by ncreated — CHANGES_REQUESTED** [review]
Great work 💪, definitely a good start. I left smaller comments - mainly on optionality of fields vs their defaults. 

Overal, I wonder on two things: 
- Shouldn't we call this format (the folder) a `mobile-session-replay` instead of `session-replay`? Both mean something different.
- Should we merge it directly to the `master` branch? I think it's fair to assume that at the beginning there will be a lot of iterations and breaking changes so maybe it's worth using feature branch for a while. On the other hand, both formats are isolated, so I'm ok with any approach.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: Well the thing is that I don't see the reason to provide defaults for them in that case. If they are optionals, their default is `null`.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: Yeah, it might be so. We can add them there.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: Like I mentioned above, in case we mark a field as optional that means that it will default to `null`. In that case we don't need special defaults for it. Would you rather go with optionals for these fields ?

File: schemas/session-replay/text-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/text-wireframe-schema.json]: Good point, I would like to use #RRGGBBAA here so I will specify it in the description.

File: schemas/session-replay/text-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/text-wireframe-schema.json:38]: > Great work 💪, definitely a good start. I left smaller comments - mainly on optionality of fields vs their defaults.
> 
> Overal, I wonder on two things:
> 
> * Shouldn't we call this format (the folder) a `mobile-session-replay` instead of `session-replay`? Both mean something different.
> * Should we merge it directly to the `master` branch? I think it's fair to assume that at the beginning there will be a lot of iterations and breaking changes so maybe it's worth using feature branch for a while. On the other hand, both formats are isolated, so I'm ok with any approach.

@ncreated 
1. Regarding the format name I would rather keep it `session-replay` as this might relate with the browser session replay later. @BenoitZugmeyer What do you think ?
2. Like you also said, they are separated so I would rather merge them on master but I am waiting for others opinion also.

File: schemas/session-replay/text-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/text-wireframe-schema.json]: We could...was following the RFC here but it might be better to make it non - optional

**Review by ThibautGeriz — COMMENTED** [review]
Did not looking into details but this is only the shapes you're introducing aka the mobile FS content `data.wireframes`, not the records themselves. Do you plan to keep it that way?

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: On the other hand I get what you're saying, we can have it optional and add the `default` property just for visibility. This way at the player level they will know to which value to default if they can't find this option.

File: schemas/session-replay/_common-wireframe-schema.json
Comment (ThibautGeriz) [inline schemas/session-replay/_common-wireframe-schema.json]: since we're talking about a shape, we should add a comment to which part of the shape we're talking about, top-left?

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ncreated) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: > Well the thing is that I don't see the reason to provide defaults for them in that case. If they are optionals, their default is null.

I don't mean providing defaults in JSON schema. I mean **implementing** defaults in the player. // cc @DataDog/rum-session-replay 

I think we mix optionality with nullability here, but these are two separate concepts. By including `opacity` as `required` parameter it will be generated as:
```swift
let opacity: Double
```
on our side. No option to set `null` (Kotlin) or `nil` (Swift).

If we don't list `opacity` in `required`, it will be optional:
```swift
let opacity: Double?
```
hence we will be able to pass `nil` making the field absent in the JSON or (in Swift) `Optional<Double>.none` making it use `null` in the JSON.


Ref.: [Required Properties](https://json-schema.org/understanding-json-schema/reference/object.html#required-properties) on json-schema.org.

File: schemas/session-replay/_common-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/_common-wireframe-schema.json]: @ThibautGeriz the idea is that I wanted to start small, I will add the records in another PR once this one is merged. Of course the records and segments will be added here.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: Ok, that's clear now. We might go to provide a "default" in the JSON schema only for visibility but I am not sure. @ThibautGeriz what do you think, how do you prefer to have the defaults specified for each optional property ? Should  this be added in the schema also ?

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: I'd rather go with the #RRGGBBAA which covers both.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ncreated) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: 👍 alternatively we could guide player's implementation by explaining each default in the description.

File: schemas/session-replay/shape-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/shape-wireframe-schema.json:16]: well, it is a constant so there's no need to mark it as required but will add it there for visibility.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ThibautGeriz) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: I think it's fine to have default in make the property optional. It's even better if we can match the browser default. For example for opacity, it's no opacity

File: schemas/session-replay/_common-wireframe-schema.json
Comment (ThibautGeriz) [inline schemas/session-replay/_common-wireframe-schema.json]: > normalised based on the device DPI.

**question:** What does that mean?

Comment (mariusc83) [reply schemas/session-replay/_common-wireframe-schema.json]: Good point, I am going to detail more in the description.

File: schemas/session-replay/text-wireframe-schema.json
Comment (acorretti) [reply schemas/session-replay/text-wireframe-schema.json:38]: > Shouldn't we call this format (the folder) a mobile-session-replay instead of session-replay? Both mean something different.

I think the directory can be called `session-replay` but then the `session-replay-events-schema.json` should reference mobile, e.g. `session-replay-events-mobile.json`. This makes room for browser ones to be added later. Wireframe definitions can keep their current names, there's little chance for collision with browser.

**Review by acorretti — CHANGES_REQUESTED** [review]

File: .github/CODEOWNERS
Comment (ThibautGeriz) [inline .github/CODEOWNERS:5]: **praise:** nice, thank you

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ThibautGeriz) [inline schemas/session-replay/_common-shape-wireframe-schema.json]: **question:** what's the unit of that: px? percentage? In the browser, we also can configure different radius for each corner, it's not possible in mobile?
for reference here is the browser configuration: https://developer.mozilla.org/en-US/docs/Web/CSS/border-radius

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ThibautGeriz) [inline schemas/session-replay/_common-shape-wireframe-schema.json]: **question:** stroke is the border? if so can you add that in the description so we have both name to help us do the mapping later on in the browser

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ThibautGeriz) [inline schemas/session-replay/_common-shape-wireframe-schema.json]: **thought:** not so sure we should have a default here, if there is a border, we should specify it's width. Well for browser it does not make sense but do mobile have mostly border of 1px? 🤔

File: schemas/session-replay/_common-wireframe-schema.json
Comment (ThibautGeriz) [inline schemas/session-replay/_common-wireframe-schema.json:39]: **praise:** thank you for clarifying
In reality we don't really care, we only care about the ratio, we'll just divide one by the other

File: schemas/session-replay/text-wireframe-schema.json
Comment (ThibautGeriz) [inline schemas/session-replay/text-wireframe-schema.json]: **question:** shouldn't that be a enum?

File: schemas/session-replay/text-wireframe-schema.json
Comment (ThibautGeriz) [inline schemas/session-replay/text-wireframe-schema.json:63]: **question:** there is no regex for string in this framework?
**suggestion:** the `AA` part should/can be optional

**Review by ThibautGeriz — COMMENTED** [review]
Looks really cool

File: samples/session-replay-mobile-schema/session-replay/shape.json
Comment (jagracey) [inline samples/session-replay-mobile-schema/session-replay/shape.json:8]: Consider naming `borderRadius` to match the browser?

File: samples/session-replay-mobile-schema/session-replay/shape.json
Comment (jagracey) [inline samples/session-replay-mobile-schema/session-replay/shape.json:9]: Consider naming `backgroundColor` to match the browser for HTML (not XML).

**Comment by jagracey** [comment]
I left two comments about naming conventions that seem to roughly follow SVG. Perhaps we should be following HTML/CSS naming conventions since IMO, we would probably not be using SVG.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ncreated) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: > Well for browser it does not make sense but do mobile have mostly border of 1px? 🤔

I wouldn't say so. IMO most commonly there are no borders/strokes and surely there is no (iOS) guideline we can refer to. That said, IMO we should rather make `width` a `required` attribute and don't reason about the default. This way the real default that player can implement is "no stroke".

File: samples/session-replay-mobile-schema/session-replay/shape.json
Comment (mariusc83) [reply samples/session-replay-mobile-schema/session-replay/shape.json:8]: I would rather keep it mobile oriented and you can do the analogy on the player side. @ncreated what do you think ?

File: samples/session-replay-mobile-schema/session-replay/shape.json
Comment (mariusc83) [reply samples/session-replay-mobile-schema/session-replay/shape.json:9]: same as above, @ncreated need your opinion here also.

File: schemas/session-replay/text-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/text-wireframe-schema.json:63]: We don't need a regex actually, but we can make the alpha optional. @ncreated thoughts ?

File: schemas/session-replay/text-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/text-wireframe-schema.json]: Good question, I identified the following 5 types:  serif, sans serif, script, monospaced, and display. I think these should be enough, what do you think ?

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: px, I will add more info into the description.

File: schemas/session-replay/text-wireframe-schema.json
Comment (ThibautGeriz) [reply schemas/session-replay/text-wireframe-schema.json]: I think let's type the field as much as we can and use an enum. If we missed one, let's add it after.

File: samples/session-replay-mobile-schema/session-replay/shape.json
Comment (ThibautGeriz) [reply samples/session-replay-mobile-schema/session-replay/shape.json:8]: I agree, let's keep it mobile oriented, just make sure to add the browser equivalent in the description for clarity

File: schemas/session-replay/text-wireframe-schema.json
Comment (ThibautGeriz) [reply schemas/session-replay/text-wireframe-schema.json:38]: I would not worry to much about that because in the end, the entry point is going to be segment with both browser & mobile datastructure

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ThibautGeriz) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: Sorry I am not sure I got the take away here, i see two cases:
- no border/stroke there nothing is the default
- there is a border/stroke therefore we need to give a width

Did I got it right?

Comment (mariusc83) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: yes that's it.

File: schemas/session-replay/text-wireframe-schema.json
Comment (ThibautGeriz) [reply schemas/session-replay/text-wireframe-schema.json:38]: To be a bit more precise, the way I see it once we documented everything, it would look something like that:
```
schema/session-replay
schema/session-replay/segment.json
schema/session-replay/record.json
schema/session-replay/meta-record.json
schema/session-replay/focus-record.json
schema/session-replay/viewport-record.json
schema/session-replay/mobile/fullsnapshot-record.json
schema/session-replay/mobile/shape-wireframe.json
schema/session-replay/mobile/text-wireframe.json
schema/session-replay/browser/fullsnapshot-record.json
schema/session-replay/browser/serialized-node.json
```
What do you think?
(this is out of scope of the current PR)

Comment (mariusc83) [reply schemas/session-replay/text-wireframe-schema.json:38]: Exactly what I was thinking also. If we all agree I can refactor to this folders architecture in the next PR where I am going to add the segment, records schemas.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (0xnm) [inline schemas/session-replay/_common-shape-wireframe-schema.json]: just curious why `rgba` over `argb` (the latter is used on Android)?

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (0xnm) [inline schemas/session-replay/_common-shape-wireframe-schema.json:25]: In theory the opacity is defined by the alpha component of the `shapeStyle#fillColor`, so we are kind of duplicating info. Or is it meant to be applied on top what if defined in the `shapeStyle#fillColor`? Say we have 0.5 here and 0.5 in the `shapeStyle#fillColor`, resulting alpha will be 0.25 or 0.5?

If it is a duplication, then it is ok for me, because a) it is not a lot of data; b) has a better clarity. Alternatively we can just keep rgb and alpha in the different fields. If it is an additional alpha, then I think we should add a notice to the description.

File: schemas/session-replay/_common-wireframe-schema.json
Comment (0xnm) [inline schemas/session-replay/_common-wireframe-schema.json]: it is missing in the `properties` below. Should it be in the `required` list here? Seems it comes only with `ShapeWireframe`.

File: schemas/session-replay/text-wireframe-schema.json
Comment (0xnm) [inline schemas/session-replay/text-wireframe-schema.json]: I think we should specify a unit here. Is it supposed to be in points?

**Review by 0xnm — COMMENTED** [review]
Nicely done! I've left some comments, questions, but overall lgtm!

File: validate.sh
Comment (BenoitZugmeyer) [reply validate.sh:10]: You could move this logic to `validate.py` and enjoy a more evolved language :)

File: schemas/session-replay/text-wireframe-schema.json
Comment (BenoitZugmeyer) [reply schemas/session-replay/text-wireframe-schema.json:38]: I like what @ThibautGeriz is suggesting. We'll consider adding browser-related schemas in the future.

Comment (ThibautGeriz) [reply schemas/session-replay/text-wireframe-schema.json:38]: We'll have to it if we want to use the typescript type generated by the this repo.

Comment (BenoitZugmeyer) [reply schemas/session-replay/text-wireframe-schema.json:38]: @ThibautGeriz can't you use both the Browser SDK types and the types from this repo at first?

File: schemas/session-replay/text-wireframe-schema.json
Comment (ncreated) [reply schemas/session-replay/text-wireframe-schema.json:63]: Both (optional and non-optional `AA`) works for me. At the end, we will be exporting the value from r, g, b, a `floats` and constructing the hex manually, so I don't think we will have any preference.

File: samples/session-replay-mobile-schema/session-replay/shape.json
Comment (ncreated) [reply samples/session-replay-mobile-schema/session-replay/shape.json:9]: Thing is, `shapeStyle` bundles more than background color: stroke width, stroke color, fill color, corner radius, opacity, ... . It could make sense to call `fillColor` a `backgroundColor`, but IMO not the `shapeStyle`.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (ncreated) [reply schemas/session-replay/_common-shape-wireframe-schema.json:25]: > Or is it meant to be applied on top what if defined in the shapeStyle#fillColor?

Yes 👍.

My original thought behind this was to separate wireframe's `opacity` from fill / stroke opacity (embedded in `#RRGGBBAA`). Otherwise, when exporting wireframes we'd need to perform blending on the SDK side to result with proper values in hex colors.

File: samples/session-replay-mobile-schema/session-replay/shape.json
Comment (jagracey) [reply samples/session-replay-mobile-schema/session-replay/shape.json:9]: Similarly there are a few more styles I was thinking of, like fill and stroke being SVG based as well.
Anyway no big deal at this point.

File: schemas/session-replay/_common-shape-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/_common-shape-wireframe-schema.json]: I think the `RRGGBBAA` is more popular

File: schemas/session-replay/_common-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/_common-wireframe-schema.json]: not it is a left over from a prev iteration ;)

File: schemas/session-replay/text-wireframe-schema.json
Comment (mariusc83) [reply schemas/session-replay/text-wireframe-schema.json:63]: We can keep it as it is but and have in mind that we can also skip the alpha if needed.

File: validate.sh
Comment (mariusc83) [reply validate.sh:10]: Yes, did not want to do this in this PR, maybe later in a dedicated PR.

File: schemas/session-replay/text-wireframe-schema.json
Comment (ThibautGeriz) [reply schemas/session-replay/text-wireframe-schema.json:38]: @BenoitZugmeyer Kinda but it's gonna be awkward, we'll start with that. As I target I wish we could avoid. The segment we fetch we'll have to cast it `MobileSegment | BrowserSegment` with both having the same definition for the focus record for example

**Review by ThibautGeriz — APPROVED** [review]

**Review by acorretti — APPROVED** [review]

**Review by ncreated — APPROVED** [review]
Well done 💪

**Pull request merged**
